apis/vpcaccess/v1beta1/connector_reference.go

```diff
@@ -0,0 +1,165 @@
+type VPCAccessConnectorID struct {
+	// Immutable.
+	// The VPCAccessConnector name. If not given, the metadata.name will be used.
+	ResourceID *string `json:"resourceID,omitempty"`
+
+	// Location represents the geographical location of the VPCAccessConnector. Specify a region name or "global" for global resources. Reference: GCP definition of regions/zones (https://cloud.google.com/compute/docs/regions-zones/)
+	Location string `json:"location"`
+
+	// The project that this resource belongs to.
+	ProjectRef *refs.ProjectRef `json:"projectRef,omitempty"`
+}
+
+func (i *VPCAccessConnectorID) String() string {
+	return fmt.Sprintf("projects/%s/locations/%s/connectors/%s", i.ProjectRef.External, i.Location, direct.ValueOf(i.ResourceID))
+}
+
+// NormalizedExternal provision the "External" value for other resource that depends on VPCAccessConnector.
+// If the "External" is given in the other resource's spec.VPCAccessConnectorRef, the given value will be used.
+// Otherwise, the "Name" and "Namespace" will be used to query the actual VPCAccessConnector object from the cluster.
+func (r *VPCAccessConnectorRef) NormalizedExternal(ctx context.Context, reader client.Reader, otherNamespace string) (string, error) {
+	if r.External != "" && r.Name != "" {
+		return "", fmt.Errorf("cannot specify both name and external on %s reference", VPCAccessConnectorGVK.Kind)
+	}
+	// Verify external External
+	if r.External != "" {
+		id, err := parseVPCAccessConnectorExternal(r.External)
+		if err != nil {
+			return "", err
+		}
+		// Store normalized form
+		r.External = id.String()
+		return r.External, nil
+	}
+
+	// From the Config Connector object
+	if r.Namespace == "" {
+		r.Namespace = otherNamespace
+	}
+	key := types.NamespacedName{Name: r.Name, Namespace: r.Namespace}
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(VPCAccessConnectorGVK)
+	if err := reader.Get(ctx, key, u); err != nil {
+		if apierrors.IsNotFound(err) {
+			return "", k8s.NewReferenceNotFoundError(u.GroupVersionKind(), key)
+		}
+		return "", fmt.Errorf("reading referenced %s %s: %w", VPCAccessConnectorGVK, key, err)
+	}
+	// Get external from status.externalRef. This is the most trustworthy place.
+	actualExternalRef, _, err := unstructured.NestedString(u.Object, "status", "externalRef")
+	if err != nil {
+		return "", fmt.Errorf("reading status.externalRef: %w", err)
+	}
+	if actualExternalRef == "" {
+		return "", k8s.NewReferenceNotReadyError(u.GroupVersionKind(), key)
+	}
+	r.External = actualExternalRef
+	return r.External, nil
+}
+
+// NewVPCAccessConnectorID builds a VPCAccessConnectorID from the Config Connector VPCAccessConnector object.
+func NewVPCAccessConnectorID(ctx context.Context, reader client.Reader, obj *VPCAccessConnector) (*VPCAccessConnectorID, error) {
+	id := &VPCAccessConnectorID{}
+
+	*id = obj.Spec.VPCAccessConnectorID
+
+	// Normalize projectRef
+	projectRef, err := refsv1beta1.ResolveProject(ctx, reader, obj, obj.Spec.ProjectRef)
```

Comment:
Looks like this resource does not have `spec.projectRef`.  The only option we have is`ResolveProjectID`. (Those `Project` lib is error-prone and confusing. it's been in my todo for a while)